File: ros2_ws/src/activity4_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"
#include <random>
#include "my_robot_interfaces/srv/set_led.hpp"

class BatteryNode : public rclcpp::Node
{
public:
    BatteryNode() : Node("battery")
    {
        std::uniform_int_distribution<int64_t> distrib_led(0, 2);
        std::uniform_int_distribution<> distrib_bool(0, 1);
        battery_state_ = 100;
        timer_ = this->create_wall_timer(std::chrono::seconds(4), std::bind(&BatteryNode::callbackWaitFourSeconds, this));
        RCLCPP_INFO(this->get_logger(), "Battery node  is running");
    }

private:
    int battery_state_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::thread thread_;

    // Random
    std::mt19937 gen;
    std::uniform_int_distribution<> distrib_led;
    std::uniform_int_distribution<> distrib_bool;

    void callbackWaitFourSeconds()
    {
        battery_state_ = 0;

        // Send the request
        thread_ = std::thread(std::bind(&BatteryNode::callbackSetLed, this));

        // Wait for 6 seconds
        std::this_thread::sleep_for(std::chrono::seconds(6));
    }

    void callbackSetLed()
    {
        auto client = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "waiting for the server to be up...");
        }
        auto request = std::make_shared<my_robot_interfaces::srv::SetLed::Request>();

        int led_number = distrib_led(gen);
        bool state = (distrib_bool(gen) == 1);

        request->led_number = led_number;
        request->state = state;

        auto future = client->async_send_request(request);
        try
        {
            auto response = future.get();
            bool success = response->success;
            if (success)
            {
                std::string status = state ? "on" : "off";
                RCLCPP_INFO(this->get_logger(), "Led %d changed to %s", led_number, status.c_str());
            }
            else
            {
                RCLCPP_INFO(this->get_logger(), "It was not posiible to change the state of %d", led_number);
            }
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call feiled");
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/activity4_py_pkg/activity4_py_pkg/battery.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.task import Future

import time
import random
from functools import partial

from my_robot_interfaces.srv import SetLed



class BattryNode(Node):
    def __init__(self):
        super().__init__("battery")
        self.battery_state_ = 100
        self.timer_empty = self.create_timer(4.0, self.drain_battery)
        self.client_ = self.create_client(SetLed, "set_led")

        while not self.client_.wait_for_service(
            1.0
        ):  # The 1.0 is to wait for service (1 second)
            self.get_logger().warn("Waiting for Server Add Two Ints...")

    def drain_battery(self):
        self.battery_state_ = 0
        # Send request to turn on led
        led_number = random.choice([0, 1, 2])
        state = random.choice([True, False])

        request = SetLed.Request()
        request.led_number = led_number
        request.state = state

        future = self.client_.call_async(request)
        future.add_done_callback(partial(self.callback_call_set_led, led_number=led_number, state=state))

        time.sleep(6)
        self.battery_state_ = 100

    def callback_call_set_led(self, future: Future, led_number, state):
        try:
            response: SetLed.Response = future.result()
            success = response.success
            if success:
                self.get_logger().info(f"Led {led_number} turned {state}")
            else:
                self.get_logger().info(f"Problem to change state from {led_number}")
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = BattryNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
